design-proposals/klip-7.md

```diff
@@ -0,0 +1,181 @@
+# Query Execution Plans 
+
+Author: Rohan Desai (rodesai) 
+
+Release target: 5.4 
+
+Status: In Discussion
+
+Discussion:
+
+## Motivation and background
+
+One of the main differences between KSQL and a database is that KSQL queries are long-running. A database query is issued by a user, executes, and completes. A KSQL query usually has no defined end time. It keeps running until a user explicitly asks KSQL to stop running it. This is challenging because KSQL needs to ensure that it correctly runs queries across software versions, even as the engine that builds execution plans changes.
+
+As an example of what can go wrong, consider the query:
+
+`CREATE TABLE AGG AS SELECT X, Y, COUNT(*) FROM SOURCE;`
+
+In today's KSQL, the key for AGG is computed by concatenating the string forms of X and Y. In the future we may wish to create a structured key instead. However, we need to be sure to continue using the concatenated string format for queries that were started before that change was made.
+
+Our current approach to this problem is to do our best not to break compatibility by toggling incompatible behavior on compatibility-breaking configs, and to test that queries are built compatibly across versions. However, there are a number of problems with this approach:
+
+- The tests have some gaps. For example, we can't always catch incompatibility from pushing up filters or reordering internal projections
+- Tests are always incomplete, and breaking compatibility can lead to errors that are unrecoverable.
+- The tests rely on an unstructured topology format taken from Kafka Streams. This is very brittle. Part of this proposal is to switch to a more structured format.
```

Comment:
As mentioned, I've added internal topics to _some_ QTT tests, see https://github.com/confluentinc/ksql/blob/6af8958f2a794e480a07916b4c778f9f51ae614f/ksql-functional-tests/src/test/resources/query-validation-tests/joins.json#L22-L29 for an example.